### DISCUSSION TITLE
Duplicated hierarchical tags

### DISCUSSION THREAD

**Discussion by rhyst** [immich-app/immich#12207] [Q&A]
### The bug

Immich is displaying each level of a hierarchical tag as a top level tag (incorrectly), as well as a (correctly) nested tag.

I have an image and with `exiftool` I can check the tags:

```
Categories                      : <Categories><Category Assigned="0">Canyon<Category Assigned="1">Birks of Aberfeldy</Category></Category><Category Assigned="0">Trip<Category Assigned="0">2024<Category Assigned="1">24.08.17 - Canyoning Birks of Aberfeldy</Category></Category></Category><Category Assigned="0">Activity<Category Assigned="1">Canyoning</Category></Category></Categories>
Tags List                       : Canyon/Birks of Aberfeldy, Trip/2024/24.08.17 - Canyoning Birks of Aberfeldy, Activity/Canyoning
Last Keyword XMP                : Canyon/Birks of Aberfeldy, Trip/2024/24.08.17 - Canyoning Birks of Aberfeldy, Activity/Canyoning
Hierarchical Subject            : Canyon|Birks of Aberfeldy, Trip|2024|24.08.17 - Canyoning Birks of Aberfeldy, Activity|Canyoning
Catalog Sets                    : Canyon|Birks of Aberfeldy, Trip|2024|24.08.17 - Canyoning Birks of Aberfeldy, Activity|Canyoning
```

If we take the `Canyon/Birks of Aberfeldy` tag as an example this appears correctly nested in the Immich UI like:

![image](https://github.com/user-attachments/assets/124f9de1-14f1-4f9b-ae2a-c92fecf7dee4)

But also the `Birks of Aberfeldy` child tag appears at the top level:

![image](https://github.com/user-attachments/assets/153c1a3e-545a-4274-be72-1aefacc3ad47)

The `Birks of Aberfeldy` tag should only appear once, nested under `Canyon`.

It's possible I have set up my tags wrong (originally tagged in Digikam) but the `exiftool` tags list seems to have them correctly "deuplicated" so I am wondering if this is an Immich issue.

### The OS that Immich Server is running on

Ubuntu 20.04.6 LTS

### Version of Immich Server

v113.0.0

### Version of Immich Mobile App

N/A

### Platform with the issue

- [X] Server
- [X] Web
- [ ] Mobile

### Your docker-compose.yml content

```YAML
I run Immich from ansible so its sort of hard to collate into one file. Let me know if you think it is relevant and I will try.
```

### Your .env content

```Shell
Same as above.
```

### Reproduction steps

1. I had photos with the tags show in the `exiftool` output above.
2. I enabled tags feature
3. I ran the "Extract Metadata" job

**Comment by jrasm91** [comment]
Can you click on the top level tag and see which assets it is linked to and then show the same output for that image?

**Reply by rhyst** [reply]
Sorry if this is not what you want, not 100% sure what you mean.

In the top level `Birks of Aberfeldy` tag I can see an image with a filename `P8171038.JPG` and and ID of `ff964df8-1d52-49eb-9183-52485dd5bfc3`. If I click the `Canyon/Birks of Aberfeldy` tag I can see the same image with same filename and ID.

In the Immich UI the tags are shown as:

![image](https://github.com/user-attachments/assets/e2f740da-803f-46f8-9649-f458ba8911ad)

And the relevant exiftool output:

```
Categories                      : <Categories><Category Assigned="0">Canyon<Category Assigned="1">Birks of Aberfeldy</Category></Category><Category Assigned="0">Trip<Category Assigned="0">2024<Category Assigned="1">24.08.17 - Canyoning Birks of Aberfeldy</Category></Category></Category><Category Assigned="0">Activity<Category Assigned="1">Canyoning</Category></Category></Categories>
Tags List                       : Canyon/Birks of Aberfeldy, Trip/2024/24.08.17 - Canyoning Birks of Aberfeldy, Activity/Canyoning
Last Keyword XMP                : Canyon/Birks of Aberfeldy, Trip/2024/24.08.17 - Canyoning Birks of Aberfeldy, Activity/Canyoning
Hierarchical Subject            : Canyon|Birks of Aberfeldy, Trip|2024|24.08.17 - Canyoning Birks of Aberfeldy, Activity|Canyoning
Catalog Sets                    : Canyon|Birks of Aberfeldy, Trip|2024|24.08.17 - Canyoning Birks of Aberfeldy, Activity|Canyoning
Subject                         : Birks of Aberfeldy, 24.08.17 - Canyoning Birks of Aberfeldy, Canyoning
```

**Reply by jrasm91** [reply]
What about `Keywords`?

**Reply by rhyst** [reply]
```
 "Keywords": ["Birks of Aberfeldy","24.08.17 - Canyoning Birks of Aberfeldy","Canyoning"],
```

My full output from other reply:

```
  "SourceFile": "/mnt/storage/Pictures/2024/24.08.17 - Canyoning Birks of Aberfeldy/P8171038.JPG",
  "ExifToolVersion": 11.88,
  "FileName": "P8171038.JPG",
  "Directory": "/mnt/storage/Pictures/2024/24.08.17 - Canyoning Birks of Aberfeldy",
  "FileSize": "5.3 MB",
  "FileModifyDate": "2024:08:21 14:38:26+00:00",
  "FileAccessDate": "2024:08:21 14:38:26+00:00",
  "FileInodeChangeDate": "2024:08:21 14:38:26+00:00",
  "FilePermissions": "rwxrwxr-x",
  "FileType": "JPEG",
  "FileTypeExtension": "jpg",
  "MIMEType": "image/jpeg",
  "ExifByteOrder": "Little-endian (Intel, II)",
  "ImageDescription": "",
  "Make": "OLYMPUS CORPORATION",
  "Model": "TG-6",
  "Orientation": "Horizontal (normal)",
  "XResolution": 314,
  "YResolution": 314,
  "ResolutionUnit": "inches",
  "Software": "Version 1.0",
  "ModifyDate": "2024:08:17 13:41:19",
  "Artist": "",
  "YCbCrPositioning": "Centered",
  "Copyright": "",
  "ExposureTime": "1/30",
  "FNumber": 2.0,
  "ExposureProgram": "Program AE",
  "ISO": 200,
  "SensitivityType": "Standard Output Sensitivity",
  "ExifVersion": "0231",
  "DateTimeOriginal": "2024:08:17 13:41:19",
  "CreateDate": "2024:08:17 13:41:19",
  "OffsetTime": "+00:00",
  "OffsetTimeOriginal": "+00:00",
  "OffsetTimeDigitized": "+00:00",
  "ComponentsConfiguration": "Y, Cb, Cr, -",
  "ExposureCompensation": 0,
  "MaxApertureValue": 2.0,
  "MeteringMode": "Multi-segment",
  "LightSource": "Unknown",
  "Flash": "Auto, Did not fire",
  "FocalLength": "4.5 mm",
  "AmbientTemperature": "11.1 C",
  "Pressure": 987,
  "WaterDepth": 0,
  "Acceleration": 997800,
  "FlashpixVersion": "0100",
  "ColorSpace": "sRGB",
  "ExifImageWidth": 4000,
  "ExifImageHeight": 3000,
  "FileSource": "Digital Camera",
  "CustomRendered": "Normal",
  "ExposureMode": "Auto",
  "WhiteBalance": "Auto",
  "DigitalZoomRatio": 1,
  "FocalLengthIn35mmFormat": "25 mm",
  "SceneCaptureType": "Standard",
  "GainControl": "Low gain up",
  "Contrast": "Normal",
  "Saturation": "Normal",
  "Sharpness": "Normal",
  "Compression": "JPEG (old-style)",
  "ThumbnailOffset": 13956,
  "ThumbnailLength": 9257,
  "XMPToolkit": "XMP Core 4.4.0-Exiv2",
  "Warning": "[minor] Fixed incorrect URI for xmlns:MicrosoftPhoto",
  "Urgency": "0 (reserved)",
  "DateTime": "2024:08:17 13:41:19",
  "MetadataDate": "2024:08:17 13:41:19",
  "PickLabel": 0,
  "ColorLabel": 0,
  "Categories": "<Categories><Category Assigned=\"0\">Canyon<Category Assigned=\"1\">Birks of Aberfeldy</Category></Category><Category Assigned=\"0\">Trip<Category Assigned=\"0\">2024<Category Assigned=\"1\">24.08.17 - Canyoning Birks of Aberfeldy</Category></Category></Category><Category Assigned=\"0\">Activity<Category Assigned=\"1\">Canyoning</Category></Category></Categories>",
  "TagsList": ["Canyon/Birks of Aberfeldy","Trip/2024/24.08.17 - Canyoning Birks of Aberfeldy","Activity/Canyoning"],
  "LastKeywordXMP": ["Canyon/Birks of Aberfeldy","Trip/2024/24.08.17 - Canyoning Birks of Aberfeldy","Activity/Canyoning"],
  "HierarchicalSubject": ["Canyon|Birks of Aberfeldy","Trip|2024|24.08.17 - Canyoning Birks of Aberfeldy","Activity|Canyoning"],
  "CatalogSets": ["Canyon|Birks of Aberfeldy","Trip|2024|24.08.17 - Canyoning Birks of Aberfeldy","Activity|Canyoning"],
  "Subject": ["Birks of Aberfeldy","24.08.17 - Canyoning Birks of Aberfeldy","Canyoning"],
  "CurrentIPTCDigest": "de45ebd305b2cc985c7515b0a383f2a0",
  "CodedCharacterSet": "UTF8",
  "DateCreated": "2024:08:17",
  "TimeCreated": "13:41:19+00:00",
  "Keywords": ["Birks of Aberfeldy","24.08.17 - Canyoning Birks of Aberfeldy","Canyoning"],
  "ImageWidth": 4000,
  "ImageHeight": 3000,
  "EncodingProcess": "Baseline DCT, Huffman coding",
  "BitsPerSample": 8,
  "ColorComponents": 3,
  "YCbCrSubSampling": "YCbCr4:2:2 (2 1)",
  "Aperture": 2.0,
  "ImageSize": "4000x3000",
  "Megapixels": 12.0,
  "ScaleFactor35efl": 5.6,
  "ShutterSpeed": "1/30",
  "SubSecCreateDate": "2024:08:17 13:41:19+00:00",
  "SubSecDateTimeOriginal": "2024:08:17 13:41:19+00:00",
  "SubSecModifyDate": "2024:08:17 13:41:19+00:00",
  "ThumbnailImage": "(Binary data 9257 bytes, use -b option to extract)",
  "DateTimeCreated": "2024:08:17 13:41:19+00:00",
  "CircleOfConfusion": "0.005 mm",
  "FOV": "71.5 deg",
  "FocalLength35efl": "4.5 mm (35 mm equivalent: 25.0 mm)",
  "HyperfocalDistance": "1.87 m",
  "LightValue": 5.9
```

**Reply by jrasm91** [reply]
Keywords are added as top level tags, since they are used a lot by other platforms for "tags". That is why it is there. We could possibly add some logic like "ignore keywords if there is a TagsList".

**Reply by rhyst** [reply]
Given that I probably need to reset things whatever happens, what is the easiest way to delete all tags from Immich (but not from the files)?

**Reply by jrasm91** [reply]
Or, like remove it from keywords if it is already a child of an item in TagsList. Basically need to change the logic to take into account that keywords is an unnested version of TagsList while still supporting it for other editors.

**Reply by jrasm91** [reply]
After we fix the issue you should run metadata extraction for all assets and then you can delete tags with 0 assets, which should be easy enough.

**Reply by rhyst** [reply]
That would work though I think I would have about 500 empty tags to delete so if there were any way to automate that it would make things easier.

**Comment by maxshcherbina** [comment]
Are you by any chance using Lightroom for your hierarchical tagging? Could be that some of the tags are not considered nested akin to my issue: #12196

**Reply by rhyst** [reply]
I used Digikam for all my tags. Your issue sounds slightly different if you are not seeing the hierarchy at all? I do see the hierarchy in the Immich but I also see all the child tags duplicated at the top level.

My structured exiftool output is:

```
  "SourceFile": "/mnt/storage/Pictures/2024/24.08.17 - Canyoning Birks of Aberfeldy/P8171038.JPG",
  "ExifToolVersion": 11.88,
  "FileName": "P8171038.JPG",
  "Directory": "/mnt/storage/Pictures/2024/24.08.17 - Canyoning Birks of Aberfeldy",
  "FileSize": "5.3 MB",
  "FileModifyDate": "2024:08:21 14:38:26+00:00",
  "FileAccessDate": "2024:08:21 14:38:26+00:00",
  "FileInodeChangeDate": "2024:08:21 14:38:26+00:00",
  "FilePermissions": "rwxrwxr-x",
  "FileType": "JPEG",
  "FileTypeExtension": "jpg",
  "MIMEType": "image/jpeg",
  "ExifByteOrder": "Little-endian (Intel, II)",
  "ImageDescription": "",
  "Make": "OLYMPUS CORPORATION",
  "Model": "TG-6",
  "Orientation": "Horizontal (normal)",
  "XResolution": 314,
  "YResolution": 314,
  "ResolutionUnit": "inches",
  "Software": "Version 1.0",
  "ModifyDate": "2024:08:17 13:41:19",
  "Artist": "",
  "YCbCrPositioning": "Centered",
  "Copyright": "",
  "ExposureTime": "1/30",
  "FNumber": 2.0,
  "ExposureProgram": "Program AE",
  "ISO": 200,
  "SensitivityType": "Standard Output Sensitivity",
  "ExifVersion": "0231",
  "DateTimeOriginal": "2024:08:17 13:41:19",
  "CreateDate": "2024:08:17 13:41:19",
  "OffsetTime": "+00:00",
  "OffsetTimeOriginal": "+00:00",
  "OffsetTimeDigitized": "+00:00",
  "ComponentsConfiguration": "Y, Cb, Cr, -",
  "ExposureCompensation": 0,
  "MaxApertureValue": 2.0,
  "MeteringMode": "Multi-segment",
  "LightSource": "Unknown",
  "Flash": "Auto, Did not fire",
  "FocalLength": "4.5 mm",
  "AmbientTemperature": "11.1 C",
  "Pressure": 987,
  "WaterDepth": 0,
  "Acceleration": 997800,
  "FlashpixVersion": "0100",
  "ColorSpace": "sRGB",
  "ExifImageWidth": 4000,
  "ExifImageHeight": 3000,
  "FileSource": "Digital Camera",
  "CustomRendered": "Normal",
  "ExposureMode": "Auto",
  "WhiteBalance": "Auto",
  "DigitalZoomRatio": 1,
  "FocalLengthIn35mmFormat": "25 mm",
  "SceneCaptureType": "Standard",
  "GainControl": "Low gain up",
  "Contrast": "Normal",
  "Saturation": "Normal",
  "Sharpness": "Normal",
  "Compression": "JPEG (old-style)",
  "ThumbnailOffset": 13956,
  "ThumbnailLength": 9257,
  "XMPToolkit": "XMP Core 4.4.0-Exiv2",
  "Warning": "[minor] Fixed incorrect URI for xmlns:MicrosoftPhoto",
  "Urgency": "0 (reserved)",
  "DateTime": "2024:08:17 13:41:19",
  "MetadataDate": "2024:08:17 13:41:19",
  "PickLabel": 0,
  "ColorLabel": 0,
  "Categories": "<Categories><Category Assigned=\"0\">Canyon<Category Assigned=\"1\">Birks of Aberfeldy</Category></Category><Category Assigned=\"0\">Trip<Category Assigned=\"0\">2024<Category Assigned=\"1\">24.08.17 - Canyoning Birks of Aberfeldy</Category></Category></Category><Category Assigned=\"0\">Activity<Category Assigned=\"1\">Canyoning</Category></Category></Categories>",
  "TagsList": ["Canyon/Birks of Aberfeldy","Trip/2024/24.08.17 - Canyoning Birks of Aberfeldy","Activity/Canyoning"],
  "LastKeywordXMP": ["Canyon/Birks of Aberfeldy","Trip/2024/24.08.17 - Canyoning Birks of Aberfeldy","Activity/Canyoning"],
  "HierarchicalSubject": ["Canyon|Birks of Aberfeldy","Trip|2024|24.08.17 - Canyoning Birks of Aberfeldy","Activity|Canyoning"],
  "CatalogSets": ["Canyon|Birks of Aberfeldy","Trip|2024|24.08.17 - Canyoning Birks of Aberfeldy","Activity|Canyoning"],
  "Subject": ["Birks of Aberfeldy","24.08.17 - Canyoning Birks of Aberfeldy","Canyoning"],
  "CurrentIPTCDigest": "de45ebd305b2cc985c7515b0a383f2a0",
  "CodedCharacterSet": "UTF8",
  "DateCreated": "2024:08:17",
  "TimeCreated": "13:41:19+00:00",
  "Keywords": ["Birks of Aberfeldy","24.08.17 - Canyoning Birks of Aberfeldy","Canyoning"],
  "ImageWidth": 4000,
  "ImageHeight": 3000,
  "EncodingProcess": "Baseline DCT, Huffman coding",
  "BitsPerSample": 8,
  "ColorComponents": 3,
  "YCbCrSubSampling": "YCbCr4:2:2 (2 1)",
  "Aperture": 2.0,
  "ImageSize": "4000x3000",
  "Megapixels": 12.0,
  "ScaleFactor35efl": 5.6,
  "ShutterSpeed": "1/30",
  "SubSecCreateDate": "2024:08:17 13:41:19+00:00",
  "SubSecDateTimeOriginal": "2024:08:17 13:41:19+00:00",
  "SubSecModifyDate": "2024:08:17 13:41:19+00:00",
  "ThumbnailImage": "(Binary data 9257 bytes, use -b option to extract)",
  "DateTimeCreated": "2024:08:17 13:41:19+00:00",
  "CircleOfConfusion": "0.005 mm",
  "FOV": "71.5 deg",
  "FocalLength35efl": "4.5 mm (35 mm equivalent: 25.0 mm)",
  "HyperfocalDistance": "1.87 m",
  "LightValue": 5.9
```

From this output I can't actually see a single field that would produce the tags list I see. `Categories`, `TagsList`, `LastKeywordXMP`, `HierarchicalSubject`, and `CatalogSets` all only contain the tags with full "paths". The `Subject` and `Keywords` tags only contain the deepest child tags. Does Immich combine multiple fields for the tags?

**Reply by maxshcherbina** [reply]
Yeah I think (not 100%) the `Subject` and `Keywords` metadata properties get read seperately from the `HierarchicalSubject` keywords and thus are doubled. For now, I think your best bet is to disable writing to the `Subject` and `Keywords` for now if that doesnt interfere with your workflow.

It seems that in the future, Immich should check if `Subject` and `Keywords` are duplicated in other taglists before adding them.

Here is how to disable writing to `Subject` and `Keywords` in digikam. Fair warning though, I don't know if it will delete these fields when you don't have them selected.
![image](https://github.com/user-attachments/assets/cf577526-fd0a-4a01-a341-adb89c4010b8)

**Reply by rhyst** [reply]
That workflow would work for me, but I think I will just wait for Immich to handle this better.

**Comment by rhyst** [comment]
@jrasm91 from the discussion above, this does still seem like a bug to me. It this discussion still the appropriate place for this, or should it be an issue?

**Comment by mattsteg** [comment]
I am also experiencing this unexpected and undesirable behavior.  I'm happy to provide example files (I personally export from Capture One and also use digikam) to verify that a solution works properly.

The typical behavior of a lot of software is to export the hierarchical keywords AND also export them as non-hierarchical to support legacy software that doesn't properly support hierarchical.  In commercial software it tends to be done silently without the ability to disable.  Extracting both and not deduplicating is generally going to give unexpected and undesired results.

I could technically export a dedicated set of images for immich to catalog, then go back and strip out the duplicate tags that I don't want read before importing...but this would be an incredible disruption of workflow.  For my working files syncing metadata back and forth the tags would be regenerated.

**Comment by jrasm91** [comment]
This should be fixed with #12286, although I might have to add some instructions/tools for cleaning up the tags that were created.

**Reply by mattsteg** [reply]
Lmao I was looking at that code and scratching my head how bare keywords were getting imported at all.

2nd time today I was looking at your tag code trying to figure out why something was happening that was already fixed.

Hierarchical Subject support would be nice too, but if  necessary I can just make sure the rest of my process is synced up and that will populate TagsList.

**Reply by maxshcherbina** [reply]
@mattsteg  Looks like #12288 adds `HierarchicalSubject` support!

**Reply by wonx** [reply]
Hey, I just updated to that version and now the tags don't show by duplicate in the photos (thanks!), but they still appear in the tag list. How can I clear those orphaned tags?

**Reply by jrasm91** [reply]
There isn't any easy way right now. You can either: 
- manually delete them one by one
- wait for us to release a job/button to prune empty tags
- SQL delete them (using the tag_closure table) - not really recommended
- use the API and write a script to loop over the tags and delete empty ones

**Reply by wonx** [reply]
I'll wait then : )

**Reply by rhyst** [reply]
What I did as a quick hack was run a script in the dev tools:

```
const response = await fetch("https://my.immich.instance/api/tags", {
    "credentials": "include",
    "method": "GET",
    "mode": "cors"
});
const tags = await response.json()
for (const id of tags .map(tag => tag.id)) {
  await fetch("https://my.immich.instance/api/tags/" + id, {
    "credentials": "include",
    "method": "DELETE",
    "mode": "cors"
  });
}
```
Doing it in firefox dev tools just meant that I didn't have to faff around with API keys or anything.

This deleted all the tags and I was able to rescan successfully with no dupes but use at your own risk.

**Reply by jrasm91** [reply]
There will be a way to manually run a clean up job in the next release.
